librad/src/net/x509.rs

```diff
@@ -0,0 +1,238 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::ops::Deref;
+
+use futures::executor::block_on;
+use keystore::sign::Signer;
+use picky_asn1::{
+    bit_string::BitString,
+    date::GeneralizedTime,
+    restricted_string::IA5String,
+    wrapper::IntegerAsn1,
+};
+use picky_asn1_der::Asn1DerError;
+use picky_asn1_x509::{
+    self as x509,
+    oids,
+    AlgorithmIdentifier,
+    DirectoryName,
+    ExtendedKeyUsage,
+    Extension,
+    Extensions,
+    GeneralName,
+    PublicKey,
+    SubjectPublicKeyInfo,
+    TBSCertificate,
+    Validity,
+    Version,
+};
+use thiserror::Error;
+
+use crate::{
+    keys,
+    peer::{self, PeerId},
+};
+
+#[derive(Debug, Error)]
+pub enum FromDerError {
+    #[error("the subject common name must be a valid PeerId")]
+    SubjectNotPeerId(#[source] peer::conversion::Error),
+
+    #[error("invalid subject public key")]
+    InvalidPublicKey,
+
+    #[error("unsupported algorithm for subject public key")]
+    UnsupportedAlgorithm,
+
+    #[error("missing subject common name")]
+    MissingSubjectCN,
+
+    #[error("subject public key and common name must be the same")]
+    KeyMismatch,
+
+    #[error(transparent)]
+    Asn1(#[from] Asn1DerError),
+}
+
+/// Self-signed X509 certificate.
+#[derive(Debug, PartialEq)]
+pub struct Certificate {
+    peer_id: PeerId,
+    cert: x509::Certificate,
+}
+
+impl Certificate {
+    /// Generate a new self-signed [`Certificate`].
+    pub fn generate<S>(signer: &S) -> Result<Self, S::Error>
+    where
+        S: Signer,
+        S::Error: std::error::Error,
+    {
+        let peer_id = PeerId::from_signer(signer);
+
+        let algorithm = AlgorithmIdentifier::new_ed25519();
+        let serial_number = IntegerAsn1::from_bytes_be_unsigned(vec![1]);
+        let issuer = DirectoryName::new_common_name(PeerId::from_signer(signer).default_encoding());
+        let subject = issuer.clone(); // self-signed
+        let subject_public_key_info = SubjectPublicKeyInfo {
+            algorithm: algorithm.clone(),
+            subject_public_key: PublicKey::Ed(
+                BitString::with_bytes(signer.public_key().as_ref()).into(),
+            ),
+        };
+        let validity = {
+            let not_before = GeneralizedTime::new(1970, 1, 1, 0, 0, 0).unwrap().into();
+            let not_after = GeneralizedTime::new(2070, 1, 1, 0, 0, 0).unwrap().into();
```

Comment:
Climate change will have taken us out by then

Reply:
This is actually a suprisingly non-trivial issue: it is _required_ to specify validity, and reasonable values would be approx. now - approx. a month or so (there is a marginal security benefit to that, as the certificate owner basically renews the proof of possessing the key). But that would mean that the network stack has to be restarted once a month.

Reply:
So we include `rad-cron-job`? :grimacing: Would be worth capturing this issue in a ticket to think about it in the long term.

Reply:
How does the latest commit fix this now? So we make the cert valid for 3 months but does that get refreshed somehow?

Reply:
Sensei Wu likes to say: never put off until tomorrow what you can do today.

The only thing I'm wondering about is: re-generating a certificate can only fail due to the signer returning an error. This is actually a rather fatal situation -- perhaps a `panic` would be appropriate then?